internal/solver/solver.go

```diff
@@ -0,0 +1,168 @@
+}
+
+// Present a challenge to the solver. The Challenge Request comes from
+// cert-manager through the webhook integration. Once presented with a challenge,
+// a DNS Record needs to be created with the challenge information so cert-manager
+// can assert that the domain is owned by user of Phonebook.
+//
+// The DNSRecord created for a challenge will have a generic "challenge" label
+// added to it. This will allow cleanup operations to get a list of all challenges
+// and remove those that needs to be cleaned up. Now, in an ideal world, the label
+// would have been specific enough so cleanup duties would be able to retrieve only
+// the proper challenge records. For example, the label could hold the ResolvedFQDN and
+// the cleanup could retrieve DNS Records by ResolvedFQDN. It is, however, impossible
+// to implement safely as Labels have a stricter validation set(1) compared to FQDN(2).
+//
+// The tradeoff is to label challenges and do client-side filtering inside the CleanUp
+// method instead. It is reasonable to think that the number of challenges at any given
+// time is pretty low, the bandwidth/CPU to filter those records on the client side seems
+// acceptable at this time.
+//
+// 1. https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set
+// 2. https://www.ietf.org/rfc/rfc1034.txt
+func (s *Solver) Present(ch *whapi.ChallengeRequest) error {
+	ctx := context.Background()
+
+	ep := &phonebook.DNSRecord{
+		ObjectMeta: meta.ObjectMeta{
+			Namespace:    ch.ResourceNamespace,
+			GenerateName: "challenge-",
+			Labels: map[string]string{
+				kChallengeLabel: kChallengeKey,
+			},
+		},
+		Spec: phonebook.DNSRecordSpec{
+			RecordType: "TXT",
+			Name:       ch.ResolvedFQDN,
+			Targets:    []string{ch.Key},
+		},
+	}
+	return s.Create(ctx, ep)
+}
+
+// Request to clean up the request after a success/failure.
+// At this point, the DNSRecord was possibly created and it needs to be
+// deleted so Phonebook can remove it from the Provider.
+//
+// As described in the Present method, records for challenges have a generic label
+// associated with them. The CleanUp method needs to retrieve all dsn records that includes
+// this label and do client side filtering to only delete records that have a matching
+// challenge Key.
+//
+// It's possible that the DNSRecord was already deleted, or that more than 1 record with the same key
+// exists. All in all, any record that has a matching label & challenge key needs to be deleted
+// when this method returns.
+//
+// Deleting the record will have Phonebook run through the finalizer and delete the record
+// on the provider's side.
+func (s *Solver) CleanUp(ch *whapi.ChallengeRequest) error {
+	ctx := context.Background()
+
+	var challenges phonebook.DNSRecordList
+
+	label, err := labels.Parse(fmt.Sprintf("%s=%s", kChallengeLabel, kChallengeKey))
+	if err != nil {
+		return fmt.Errorf("PB#0004: failed to parse the label selector -- %w", err)
```

Comment:
Do we want to update this to a dedicated section of codes for the solver? Like the PB-SLV you've got below :)